### PR TITLE
ExperimentalScanner: Catch `DownloadException` when creating archives

### PR DESCRIPTION
**Pull request by bs-ondem**
The `ExperimentalScanner` downloads missing archives for packages and whenever,
the `Downloader` fails and throws a `DownloadException` the scanner stopped
the execution. This commit moves the downloader call in a runCatching block
and creates an `OrtIssue` on failure.

Signed-off-by: Onur Demirci <onur.demirci@bosch.io>

Please ensure that your pull request adheres to our [contribution guidelines](https://github.com/oss-review-toolkit/.github/blob/main/CONTRIBUTING.md). Thank you!

### CODE REVIEW

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/experimental/ExperimentalScanner.kt [hunk 1]
```diff
@@ -145,7 +145,7 @@ class ExperimentalScanner(
         runProvenanceScanners(controller, context)
         runPathScanners(controller, context)
 
-        createMissingArchives(controller.getNestedProvenancesByPackage())
+        createMissingArchives(controller)
```
[the remark this answers is not in the document]
Comment (bs-ondem) [reply scanner/src/main/kotlin/experimental/ExperimentalScanner.kt:148]: Yes, thank you. My commit msg was probably a little bit confusing.

**Review by sschuberth — DISMISSED** [review]

File: scanner/src/main/kotlin/experimental/ExperimentalScanner.kt [hunk 2]
```diff
@@ -541,17 +541,28 @@ class ExperimentalScanner(
...
-                archiver.archive(dir, nestedProvenance.root)
-                dir.safeDeleteRecursively(force = true)
+            provenancesWithMissingArchives.forEach { (pkg, nestedProvenance) ->
+                runCatching {
+                    downloadRecursively(nestedProvenance)
+                }.onSuccess { dir ->
+                    archiver.archive(dir, nestedProvenance.root)
+                    dir.safeDeleteRecursively(force = true)
+                }.onFailure {
+                    controller.addProvenanceResolutionIssue(
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/experimental/ExperimentalScanner.kt:557]: Provenance resolution issues happen only for packages which cannot be scanned. Therefore the implementation expects that no scan results exist for those packages and later on creates empty scan results with only these issues.
Using this function here would result in duplicate scan results for those packages.

Instead, I would recommend to the the folowing:
* Add a new `addIssue()` function to the `ScanController` backed by its own map.
* Add these issues to the scan results created in `getNestedScanResultsByPackage()`, probably best by accessing above map in `buildNestedProvenanceScanResult()`.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

**Review by mnonnenmacher — APPROVED** [review]

File: scanner/src/main/kotlin/experimental/ScanController.kt [hunk 3]
```diff
@@ -58,6 +58,11 @@ class ScanController(
      */
     private val provenanceResolutionIssues = mutableMapOf<Identifier, MutableList<OrtIssue>>()
 
+    /**
+     * TODO: Docs
```
Comment (sschuberth) [inline scanner/src/main/kotlin/experimental/ScanController.kt:62]: Uhm, shouldn't this have been filled before merging?

Comment (bs-ondem) [reply scanner/src/main/kotlin/experimental/ScanController.kt:62]: My bad, I forgot to add the docs -> https://github.com/oss-review-toolkit/ort/pull/5769

**Pull request merged**